build/ArtifactBuilder/Artifacts/ZipArchive.cs

```diff
@@ -56,73 +55,73 @@ protected override void InternalBuild()
-            System.IO.Compression.ZipFile.ExtractToDirectory(_zipFilePath, unpackDir);
-            return unpackDir;
+            var unpackedDir = Path.Join(OutputDirectory, "unpacked");
+            if (Directory.Exists(unpackedDir))
+            {
+                FileHelpers.DeleteDirectories(unpackedDir);
+            }
+            System.IO.Compression.ZipFile.ExtractToDirectory(_zipFilePath, unpackedDir);
+            return unpackedDir;
         }
 
         private void ValidateContent()
         {
-            var unpackDir = Unpack();
+            var unpackedLocation = Unpack();
 
-            // framework agent
+            var installedFilesRoot = unpackedLocation;
 
-            var frameworkSourceHomeBuilderPath = $@"{_frameworkAgentComponents.HomeRootPath}\newrelichome_{_frameworkAgentComponents.Platform}";
+            var expectedComponents = GetExpectedComponents(installedFilesRoot);
 
-            var rootInstallDirComponents = RemoveSourceHomeBuilderPath(_frameworkAgentComponents.RootInstallDirectoryComponents, frameworkSourceHomeBuilderPath);
-            var frameworkAgentHomeDirComponents = RemoveSourceHomeBuilderPath(_frameworkAgentComponents.AgentHomeDirComponents, frameworkSourceHomeBuilderPath);
-            var frameworkExtensionDirComponents = RemoveSourceHomeBuilderPath(_frameworkAgentComponents.ExtensionDirectoryComponents, frameworkSourceHomeBuilderPath);
-            var frameworkWrapperXmlFiles = RemoveSourceHomeBuilderPath(_frameworkAgentComponents.WrapperXmlFiles, frameworkSourceHomeBuilderPath);
+            var unpackedComponents = ValidationHelpers.GetUnpackedComponents(installedFilesRoot);
 
-            VerifyComponentsExist(rootInstallDirComponents, unpackDir, null);
-            VerifyComponentsExist(frameworkAgentHomeDirComponents, unpackDir, FrameworkSubDirectoryName);
-            VerifyComponentsExist(frameworkExtensionDirComponents, unpackDir, FrameworkSubDirectoryName);
-            VerifyComponentsExist(frameworkWrapperXmlFiles, unpackDir, FrameworkSubDirectoryName);
+            var missingExpectedComponents = new SortedSet<string>(expectedComponents, StringComparer.OrdinalIgnoreCase);
+            missingExpectedComponents.ExceptWith(unpackedComponents);
+            foreach (var missingComponent in missingExpectedComponents)
+            {
+                throw new PackagingException($"The unpacked ZIP archive was missing the expected component {missingComponent}");
+            }
 
-            // core agent
+            var unexpectedUnpackedComponents = new SortedSet<string>(unpackedComponents, StringComparer.OrdinalIgnoreCase);
+            unexpectedUnpackedComponents.ExceptWith(expectedComponents);
+            foreach (var unexpectedComponent in unexpectedUnpackedComponents)
+            {
+                throw new PackagingException($"The unpacked ZIP archive contained an unexpected component {unexpectedComponent}");
+            }
```

Comment:
We might be able to extract this code into the helper too, if we can parameterize the part of the exception messages.

Reply:
Good idea.